### PR TITLE
Homework 04

### PR DESCRIPTION
**Pull request by Fle9ma**
# Assignment
In the previous homework, you’ve implemented a couple of endpoints for car management and used in-memory data structures as a storage.
In this homework, you will replace those in-memory data structures with PostgreSQL database.
For start, use Docker or Docker Compose (as shown during the lecture) to create local instance of PostgreSQL database. You can use whichever port you want. Also, make sure you use volumes to keep your data safe.

Verify that instance is accessible by using one of listed methods:

- `psql` tool (comes with the PostgreSQL installation package)
- DBeaver
- pgAdmin
- `docker exec` command

When you are sure database is accessible, you can start replacing in-memory repository implementations with a proper database. For this, you will need a few new dependencies as shown during the lecture:

- JDBC starter
- PostgreSQL driver
- Flyway
- Testcontainers with PostgreSQL support

With that in place, you will need to configure your application with database connection parameters and add migration script that will create empty tables that will replace in-memory maps you currently have in your application.
Now you can replace in-memory storage in your repositories with database storage and, when you are done, make sure application still works as before.
For testing, you will have to set up Testcontainers as shown during the lecture.

Write a couple of tests that will verify correctness of your repository layer.

Write a couple of integration tests that will call your endpoints and verify results without mocking anything.

# Optional
To make unit and integration test run separately, you can move integration tests in a new source set using this Gradle plugin. Running `bootJar` should also run tests in this new source set.
Using `testImplementation("<dependency_name>")` will no longer add dependencies for your new source set. You will have to figure out how to make it work (or ask us for help.)

### CODE REVIEW

File: academy_project/src/main/resources/db/migration/V1__initial_db_setup.sql [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    vin             text unique
+);
+
+CREATE TABLE car_check_up
+(
+    id           text primary key,
+    performed_at timestamp,
+    worker_name  text,
+    price        decimal,
+    car_id       text
```
Comment (ivanmiklec) [inline academy_project/src/main/resources/db/migration/V1__initial_db_setup.sql:17]: Ovdje čak možeš iskoristit `references` kako bi se referencirao na `car(id)`. Tako možeš osigurat da aplikacija baci exception na db razini ukoliko `car` ne postoji

File: academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/controller/CarController.kt [hunk 2]
```diff
@@ -35,6 +37,12 @@ class CarController(private val carService: CarService) {
         return ResponseEntity(errorMessage, HttpStatus.NOT_FOUND)
     }
 
+    @ExceptionHandler
+    fun handleCarAlreadyExistsException(exception: CarAlreadyExistsException): ResponseEntity<ErrorMessage> {
+        val errorMessage = ErrorMessage(exception.message, HttpStatus.NOT_FOUND.value())
```
Comment (ivanmiklec) [inline academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/controller/CarController.kt:42]: Pazi samo da uskladiš HTTP kod iz bodyja i headera 😄. Ovdje je Bad Request prikladniji

File: academy_project/src/main/kotlin/com/infinum/course/car/checkup/carCheckUp/service/CarCheckUpService.kt [hunk 3]
```diff
@@ -1,26 +1,29 @@
...
+            carId = carCheckUpDataDTO.carId
         )
-        carRepository.addCarCheckUp(newCheckUp)
-        repository.insert(newCheckUp)
+        carRepository.findById(carCheckUpDataDTO.carId)
+        carCheckUpRepository.insert(newCheckUp)
         return newCheckUp
     }
 
     fun getAnalytics(): MutableMap<String, Int> {
```
Comment (ivanmiklec) [inline academy_project/src/main/kotlin/com/infinum/course/car/checkup/carCheckUp/service/CarCheckUpService.kt:25]: Ovo možeš čak optimizirat (hint: sve podatke koji ti trebaju možeš dohvatit sa samo jednim upitom, bez da dohvatiš sve 😄)

**Comment by ivanmiklec** [comment]
Pogledao, izgleda super (još samo sitni komentar: imaš ime direktorija `carCheckUp`, u Javi je [standard](https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html#:~:text=Naming%20Conventions,a%20programmer%20at%20example.com%20.) imenovat pakete sa svim malim slovima, bez odvajanja riječi, onda bi to bilo `carcheckup`)

File: academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/repository/CarRepository.kt [hunk 4]
```diff
@@ -2,35 +2,105 @@ package com.infinum.course.car.checkup.car.repository
...
+        ) { resultSet, _ ->
+            Car(
+                id = resultSet.getString("id"),
+                dateAdded = LocalDate.parse(resultSet.getDate("date_added").toString()),
+                manufacturer = resultSet.getString("manufacturer"),
+                model = resultSet.getString("model"),
+                productionYear = Year.of(resultSet.getInt("production_year")),
+                vin = resultSet.getString("vin")
+            )
+        } ?: throw CarNotFoundException(id)
```
Comment (ivanmiklec) [inline academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/repository/CarRepository.kt:59]: S obzirom da je sad implementiran JDBC, u slučaju da `queryForObject` ne vrati rezultat, odmah će se bacit `org.springframework.dao.EmptyResultDataAccessException`, pa će klijent dobit 500 internal server error.

Ovdje možeš onda ili try/catchat  `EmptyResultDataAccessException` i bacit `CarNotFoundException`, ili handleat `EmptyResultDataAccessException` u `@ExceptionHandler`

File: academy_project/src/main/kotlin/com/infinum/course/car/checkup/carCheckUp/repository/CarCheckUpRepository.kt [hunk 5]
```diff
@@ -4,27 +4,65 @@ import com.infinum.course.car.checkup.carCheckUp.entity.CarCheckUp
...
+            mapOf("id" to id)
+        ) { resultSet, _ ->
+            CarCheckUp(
+                id = resultSet.getString("id"),
+                performedAt = resultSet.getTimestamp("performed_at").toLocalDateTime(),
+                workerName = resultSet.getString("worker_name"),
+                price = resultSet.getBigDecimal("price"),
+                carId = resultSet.getString("car_id")
+            )
+        } ?: throw CarCheckUpNotFoundException(id)
```
Comment (ivanmiklec) [inline academy_project/src/main/kotlin/com/infinum/course/car/checkup/carCheckUp/repository/CarCheckUpRepository.kt:40]: Isti komentar vezano uz `EmptyResultDataAccessException`

**Comment by ivanmiklec** [comment]
Još jedan sitni komentar: `/car/{id}` ruta ne vraća `carCheckUp` informacije. Čini mi se da tu informaciju ne dodaješ u DTO u sklopu `CarService.findCarById` metode

File: academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/controller/CarController.kt (hunk 2 above)
Comment (Fle9ma) [reply academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/controller/CarController.kt:42]: Promaknuo mi opet sync bodya i headera 🙈 Ty 🙌

File: academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/repository/CarRepository.kt [hunk 6]
```diff
@@ -1,36 +1,91 @@
...
+                mapOf(
+                    "id" to entry.id,
+                    "date_added" to entry.dateAdded,
+                    "manufacturer" to entry.manufacturer,
+                    "model" to entry.model,
+                    "production_year" to entry.productionYear.toString(),
+                    "vin" to entry.vin
+                )
+            )
+        } catch (e: Exception) {
```
Comment (ivanmiklec) [inline academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/repository/CarRepository.kt:32]: Samo jedan savjet, bolje je hvatat specifičan exception (u ovom slučaju `EmptyResultDataAccessException`) jer da aplikacija izgubi konekciju prema bazi i baci drugi exception, onda je prikladnije vratit 5xx status kod (to je van scopea sad 😄 , ali dajem čisto kao primjer).

**Review by ivanmiklec — APPROVED** [review]

**Pull request merged**